man/scan.man
.TH SCAN %manext1% 2014-01-20 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
scan \- produce a summary listing of nmh messages
.SH SYNOPSIS
.HP 5
.na
.B scan
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-clear " | " \-noclear ]
.RB [ \-form
.IR formatfile ]
.RB [ \-format
.IR string ]
.RB [ \-header " | " \-noheader ]
.RB [ \-width
.IR columns ]
.RB [ \-reverse " | " \-noreverse ]
.RB [ \-file
.IR filename ]
.ad
.SH DESCRIPTION
.B scan
produces a one-line-per-message listing of the specified
folder or messages.  Each
.B scan
line contains the message number
(name), the date, the \*(lqFrom:\*(rq field, the \*(lqSubject\*(rq field,
and, if room allows, some of the body of the message.  For example:
.PP
.RS 5
.nf
.ta \w'15+- 'u +\w'07/\|05x 'u +\w'Dcrocker  'u
15+	10/\|05 crocker	nned\0\0<<Last week I asked some of
16\-	10/\|05 crocker	message id format\0\0<<I recommend
18	10/\|06 brien	Re: Exit status from mkdir
19	10/\|07*brien	\*(lqscan\*(rq listing format in nmh
.fi
.RE
.PP
The `+' on message 15 indicates that it is the current message.
.PP
The `\-' on message 16 indicates that it has been replied to, as indicated
by a \*(lqReplied:\*(rq component (produced by the
.B \-annotate
switch
to the
.B repl
command).
.PP
The `*' on message 19 indicates that no \*(lqDate:\*(rq header was
present.  The time of last modification of the message is given instead.
.PP
If there is sufficient room left on the
.B scan
line after the
subject, the line will be filled with text from the body, preceded by
\*(lq<<\*(rq, and terminated by \*(lq>>\*(rq if the body is sufficiently short.
.B scan
actually reads each of the specified messages and parses them to extract
the desired fields.  During parsing, appropriate error messages will be
produced if there are format errors in any of the messages.
.PP
By default,
.B scan
will decode RFC 2047 (MIME) encoding in
these scan listings.
.B scan
will only decode these fields if your
terminal can natively display the character set used in the encoding.
You should set the appropriate
.IR locale (1)
environment variables to your native
character set, if it is not US-ASCII.  See
.IR locale (1)
for more details on the appropriate environment variables.
.PP
The switch
.BR \-reverse ,
makes
.B scan
list the messages in reverse
order.
.PP
The
.B \-file
.I filename
switch allows the user to obtain a
.B scan
listing of a mail drop file as produced by
.BR packf .
This listing
includes every message in the file (you can't scan individual messages).
The switch
.B \-reverse
is ignored with this option.
.PP
The switch
.B \-width
.I columns
may be used to specify the width of
the scan line.  The default is to use the width of the terminal.
.PP
The
.B \-header
switch produces a header line prior to the
.B scan
listing.  Currently, the name of the folder and the current date and
time are output (see the
.B HISTORY
section for more information).
.PP
If the
.B \-clear
switch is used and
.BR scan 's
output is directed
to a terminal, then
.B scan
will consult the environment variables
.B $TERM
and
.B $TERMCAP
to determine your terminal type in order
to find out how to clear the screen prior to exiting.  If the
.B \-clear
switch is used and
.BR scan 's
output is not directed to a terminal
(e.g., a pipe or a file), then
.B scan
will send a formfeed prior
to exiting.
.PP
For example, the command:
.PP
.RS 5
(scan \-clear \-header; show all \-show pr \-f) | lpr
.RE
.PP
produces a scan listing of the current folder, followed by a formfeed,
followed by a formatted listing of all messages in the folder, one
per page.  Omitting
.RB \*(lq "\-show\ pr\ \-f" \*(rq
will cause the messages to be
concatenated, separated by a one-line header and two blank lines.
.PP
To override the output format used by
.BR scan ,
the
.B \-format
.I string
or
.B \-form
.I file
switches are used.  This permits individual fields of
the scan listing to be extracted with ease.  The string is simply a format
string and the file is simply a format file.  See
.IR mh\-format (5)
for the details.
.PP
In addition to the standard
.IR mh\-format (5)
escapes,
.B scan
also recognizes the following additional
.I component
escapes:
.PP
.RS 5
.nf
.ta \w'Dtimenow  'u +\w'Returns  'u
.I "Escape	Returns	Description"
body	string	the (compressed) first part of the body
dtimenow	date	the current date
folder	string	the name of the current folder
.fi
.RE
.PP
.RB { body }
returns a string without MIME decoding,
i.e.\& the MIME boundary seperator may be the first text shown.
.PP
If no date header is present in the message, the
.I function
escapes
which operate on
.RB { date }
will return values for the date of last
modification of the message file itself.  This feature is handy for
scanning a draft folder, as message drafts usually aren't allowed
to have dates in them.
.PP
The
.B %nmhetcdir%
directory contains several format files as examples of customized
.B scan
output.
.PP
.B scan
will update the
.B nmh
context prior to starting the listing,
so interrupting a long
.B scan
listing preserves the new context.
.B nmh
purists hate this idea.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Alternate\-Mailboxes:~^To determine the user's mailboxes
^Current\-Folder:~^To find the default current folder
.fi
.SH "SEE ALSO"
.IR pick (1),
.IR show (1),
.IR mh\-format (5)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msgs "' defaults to all"
.RB ` \-format "' defaulted as described above"
.RB ` \-noheader '
.RB ` \-width "' defaulted to the width of the terminal"
.fi
.SH CONTEXT
If a folder is given, it will become the current folder.
.SH HISTORY
Prior to using the format string mechanism,
.B \-header
used to generate
a heading saying what each column in the listing was.  Format strings
prevent this from happening.
.SH BUGS
The value of each
.I component
escape is set by
.B scan
to the
contents of the first message header
.B scan
encounters with the
corresponding component name; any following headers with the same
component name are ignored.


NEWS
#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/scan.c
/* scan.c -- display a one-line "scan" listing of folder or messages
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/fmt_scan.h>
#include <h/scansbr.h>
#include <h/tws.h>
#include <h/mts.h>
#include <h/utils.h>
#include "../sbr/m_maildir.h"

#define SCAN_SWITCHES \
    X("clear", 0, CLRSW) \
    X("noclear", 0, NCLRSW) \
    X("form formatfile", 0, FORMSW) \
    X("format string", 5, FMTSW) \
    X("header", 0, HEADSW) \
    X("noheader", 0, NHEADSW) \
    X("width columns", 0, WIDTHSW) \
    X("reverse", 0, REVSW) \
    X("noreverse", 0, NREVSW) \
    X("file file", 4, FILESW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(SCAN);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(SCAN, switches);
#undef X


int
main (int argc, char **argv)
{
    int clearflag = 0, hdrflag = 0, ontty;
    int width = -1, revflag = 0;
    int i, state, msgnum;
    ivector_t seqnum = ivector_create (0);
    int unseen, num_unseen_seq = 0;
    char *cp, *maildir, *file = NULL, *folder = NULL;
    char *form = NULL, *format = NULL, buf[BUFSIZ];
    char **argp, *nfs, **arguments;
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs *mp;
    FILE *in;

    if (nmh_init(argv[0], 1)) { return 1; }

    mts_init ();
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case CLRSW: 
		    clearflag++;
		    continue;
		case NCLRSW: 
		    clearflag = 0;
		    continue;

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    format = NULL;
		    continue;
		case FMTSW: 
		    if (!(format = *argp++) || *format == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    form = NULL;
		    continue;

		case HEADSW: 
		    hdrflag++;
		    continue;
		case NHEADSW: 
		    hdrflag = 0;
		    continue;

		case WIDTHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    width = atoi (cp);
		    continue;
		case REVSW:
		    revflag++;
		    continue;
		case NREVSW:
		    revflag = 0;
		    continue;

		case FILESW:
		    if (!(cp = *argp++) || (cp[0] == '-' && cp[1]))
			adios (NULL, "missing argument to %s", argp[-2]);
		    if (strcmp (file = cp, "-"))
			file = path (cp, TFILE);
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
            folder = pluspath (cp);
	} else
            app_msgarg(&msgs, cp);
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * Get new format string.  Must be before chdir().
     */
    nfs = new_fs (form, format, FORMAT);

    /*
     * We are scanning a maildrop file
     */
    if (file) {
	if (msgs.size)
	    adios (NULL, "\"msgs\" not allowed with -file");
	if (folder)
	    adios (NULL, "\"+folder\" not allowed with -file");

	/* check if "file" is really stdin */
	if (strcmp (file, "-") == 0) {
	    in = stdin;
	    file = "stdin";
	} else {
	    if ((in = fopen (file, "r")) == NULL)
		adios (file, "unable to open");
	}

	if (hdrflag) {
	    printf ("FOLDER %s\t%s\n", file, dtimenow (1));
	}

	scan_detect_mbox_style (in);
	for (msgnum = 1; ; ++msgnum) {
	    charstring_t scanl = NULL;

	    state = scan (in, msgnum, -1, nfs, width, 0, 0,
			  hdrflag ? file : NULL, 0L, 1, &scanl);
	    charstring_free (scanl);
	    if (state != SCNMSG && state != SCNENC)
		break;
	}
	scan_finished ();
	fclose (in);
	done (0);
    }

    /*
     * We are scanning a folder
     */

    if (!msgs.size)
	app_msgarg(&msgs, "all");
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done(1);
    seq_setprev (mp);			/* set the Previous-Sequence */

    context_replace (pfolder, folder);	/* update current folder         */
    seq_save (mp);			/* synchronize message sequences */
    context_save ();			/* save the context file         */

    /*
     * Get the sequence number for each sequence
     * specified by Unseen-Sequence
     */
    if ((cp = context_find (usequence)) && *cp) {
	char **ap, *dp;

	dp = mh_xstrdup(cp);
	ap = brkstring (dp, " ", "\n");
	for (i = 0; ap && *ap; i++, ap++)
	    ivector_push_back (seqnum, seq_getnum (mp, *ap));

	num_unseen_seq = i;
        mh_xfree(dp);
    }

    ontty = isatty (fileno (stdout));

    for (msgnum = revflag ? mp->hghsel : mp->lowsel;
	 (revflag ? msgnum >= mp->lowsel : msgnum <= mp->hghsel);
	 msgnum += (revflag ? -1 : 1)) {
	if (is_selected(mp, msgnum)) {
	    charstring_t scanl = NULL;

	    if ((in = fopen (cp = m_name (msgnum), "r")) == NULL) {
		    admonish (cp, "unable to open message");
		continue;
	    }

	    if (hdrflag) {
		printf ("FOLDER %s\t%s\n", folder, dtimenow(1));
	    }

	    /*
	     * Check if message is in any sequence given
	     * by Unseen-Sequence profile entry.
	     */
	    unseen = 0;
	    for (i = 0; i < num_unseen_seq; i++) {
		if (in_sequence(mp, ivector_at (seqnum, i), msgnum)) {
		    unseen = 1;
		    break;
		}
	    }

	    switch (state = scan (in, msgnum, 0, nfs, width,
			msgnum == mp->curmsg, unseen,
			folder, 0L, 1, &scanl)) {
		case SCNMSG: 
		case SCNENC: 
		case SCNERR: 
		    break;

		default: 
		    adios (NULL, "scan() botch (%d)", state);

		case SCNEOF: 
		    inform("message %d: empty", msgnum);
		    break;
	    }
	    charstring_free (scanl);
	    scan_finished ();
	    hdrflag = 0;
	    fclose (in);
	    if (ontty)
		fflush (stdout);
	}
    }

    ivector_free (seqnum);
    folder_free (mp);	/* free folder/message structure */
    if (clearflag)
	nmh_clear_screen ();

    done (0);
    return 1;
}
